contracts/rewards/StableTRIStaking.sol

```diff
@@ -405,14 +369,94 @@ contract StableTRIStaking is Ownable, ERC20 {
+        UserInfo storage userSender = userInfo[_sender];
+        uint256 _previousAmountSender = userSender.amount;
+        require(_amount <= _previousAmountSender, "StableTRIStaking: withdraw amount exceeds balance");
+        uint256 _newAmountSender = userSender.amount.sub(_amount);
+        userSender.amount = _newAmountSender;
+        _harvest(_sender, _previousAmountSender, _newAmountSender);
+    }
+
+    /**
+     * @notice Internal function called before an address receives pTRI tokens.
+     * Function manages UserInfo and distributes any outstanding rewards
+     * @param _recipient The address of the user which is receiving pTRI tokens
+     * @param _amount The amount of the tokens being received
+     */
+    function _beforeReceive(address _recipient, uint256 _amount) internal {
+        // Managing userInfo + pendingTokens of the receiver
+        UserInfo storage userRecepient = userInfo[_recipient];
+        uint256 _previousAmountRecepient = userRecepient.amount;
+        uint256 _newAmountRecepient = userRecepient.amount.add(_amount);
+        userRecepient.amount = _newAmountRecepient;
+        _harvest(_recipient, _previousAmountRecepient, _newAmountRecepient);
+    }
+
+    /**
+     * @notice external harvest function to claim only rewards till this block
+     */
+    function harvest() external {
+        UserInfo storage user = userInfo[_msgSender()];
+        _harvest(_msgSender(), user.amount, user.amount);
+    }
+
+    /**
+     * @dev See {IERC20-transfer}.
+     *
+     * Requirements:
+     *
+     * - `recipient` cannot be the zero address.
+     * - the caller must have a balance qnd userInfo.amount of at least `amount`.
+     */
+    function transfer(address recipient, uint256 amount) public virtual override returns (bool) {
+        // manage user info and pendingRewards
+        _beforeSend(_msgSender(), amount);
+        _beforeReceive(recipient, amount);
+
+        _transfer(_msgSender(), recipient, amount);
+        return true;
+    }
+
+    /**
+     * @dev See {IERC20-transferFrom}.
+     *
+     * Emits an {Approval} event indicating the updated allowance. This is not
+     * required by the EIP. See the note at the beginning of {ERC20}.
+     *
+     * Requirements:
+     *
+     * - `sender` and `recipient` cannot be the zero address.
+     * - `sender` must have a balance of at least `amount`.
+     * - the caller must have allowance for ``sender``'s tokens of at least
+     * `amount`.
+     */
+    function transferFrom(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) public virtual override returns (bool) {
+        // manage user info and pendingRewards
+        _beforeSend(sender, amount);
+        _beforeReceive(recipient, amount);
+
+        _transfer(sender, recipient, amount);
+        uint256 newAllowance = allowance(sender, _msgSender()).sub(amount, "ERC20: transfer amount exceeds allowance");
```

Comment:
@rej156 do you think we should put the approve before the transfer?
is there any chance of reentry?

Reply:
ideally no since we are not giving control to any other contract, but just want to be sure

Reply:
would say no, assuming it would fail otherwise